code/datums/components/storage.dm

```diff
@@ -0,0 +1,335 @@
+	if (src.max_wclass <= I.w_class)
+		return COMSIG_RETURN_FAILURE
+	return COMSIG_RETURN_SUCCESS
+
+/datum/component/storage/proc/emp_act(atom/source)
+	if (source.contents.len)
+		for (var/atom/A in source.contents)
+			if (isitem(A))
+				var/obj/item/I = A
+				I.emp_act()
+				SEND_SIGNAL(I, COMSIG_MOVABLE_EMP_ACT)
+	return
+
+// convenience function
+/datum/component/storage/proc/get_contents_len(atom/source)
+	var/items = 0
+	for (var/obj/item/I in source.contents)
+		if(!istype(I, /obj/item/grab))
+			items++
+	return items
+
+// let user add I to source; does not have checks and should be internal
+/datum/component/storage/proc/add_item(atom/source, obj/item/I, mob/user)
+	user.u_equip(I)
+	I.dropped(user)
+	I.set_loc(source)
+	hud.add_item(I)
+	return COMSIG_RETURN_SUCCESS
+
+// handles user opening the source and displaying the hud
+/datum/component/storage/proc/handle_storage(mob/user)
+	var/atom/A = src.parent
+	if (!src.sneaky)
+		playsound(A.loc, "rustle", 50, 1, -2)
+	if (A.loc == user && (!src.does_not_open_in_pocket || A == user.l_hand || A == user.r_hand))
+		if (ishuman(user))
+			var/mob/living/carbon/human/H = user
+			if (H.limbs)
+				if ((A == H.l_hand && istype(H.limbs.l_arm, /obj/item/parts/human_parts/arm/left/item)) || (A == H.r_hand && istype(H.limbs.r_arm, /obj/item/parts/human_parts/arm/right/item)))
+					return
+		if (user.s_active)
+			user.detach_hud(user.s_active)
+			user.s_active = null
+		if (src.mousetrap_check(user))
+			return
+		user.s_active = hud
+		hud.update()
+		user.attach_hud(hud)
+		A.add_fingerprint(user)
+		animate_storage_rustle(A)
+		return COMSIG_RETURN_EARLY
+	else
+		for (var/mob/M in hud.mobs)
+			if (M != user)
+				M.detach_hud(hud)
+		hud.update()
+
+// run once on Initialize()
+/datum/component/storage/proc/make_my_stuff(list/spawn_contents)
+	for (var/thing in spawn_contents)
+		var/amt = 1
+		if (!ispath(thing))
+			continue
+		if (isnum(spawn_contents[thing]))
+			amt = abs(spawn_contents[thing])
+		for (amt, amt>0, amt--)
+			new thing(src.parent)
+	return 1
+
+/datum/component/storage/proc/mousetrap_check(mob/user)
+	var/atom/A = src.parent
+	if (!ishuman(user) || user.stat)
+		return
+	for (var/obj/item/mousetrap/MT in A.contents)
+		if (MT.armed)
+			user.visible_message("<span class='alert'><B>[user] reaches into [src.parent] and sets off a mousetrap!</B></span>",\
+			"<span class='alert'><B>You reach into [src.parent], but there was a live mousetrap in there!</B></span>")
+			MT.triggered(user, user.hand ? "l_hand" : "r_hand")
+			. = 1
+	for (var/obj/item/mine/M in A.contents)
+		if (M.armed && M.used_up != 1)
+			user.visible_message("<span class='alert'><B>[user] reaches into [src.parent] and sets off a [M]!</B></span>",\
+			"<span class='alert'><B>You reach into [src.parent], but there was a live [M] in there!</B></span>")
+			M.triggered(user)
+			. = 1
+
+// future: refactor into /datum/component/storage/shared
```

Comment:
Advice + recommendations on the best way to do this appreciated.

Reply:
That will cause a lot of problems. I'd probably go with something like a global associative list `shared_storages` and in `Initialize` of this component you'd pass it which shared storage should it use (e.g. it would work with `shared_storages["bible"]`).